File: metrics/descriptors/src/main/kotlin/org/convos/metrics/descriptors/MetricsStableIdEncoder.kt
package org.convos.metrics.descriptors

import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

class MetricsStableIdEncoder(
    private val salt: ByteArray,
    private val info: ByteArray,
) {
    fun derive(privateKey: ByteArray): String {
        val prk = hmacSha256(if (salt.isEmpty()) ByteArray(SHA256_LEN) else salt, privateKey)
        val t = hmacSha256(prk, info + byteArrayOf(0x01))
        return t.joinToString("") { "%02x".format(it) }
    }

    private fun hmacSha256(key: ByteArray, data: ByteArray): ByteArray {
        val mac = Mac.getInstance(HMAC_ALG)
        mac.init(SecretKeySpec(key, HMAC_ALG))
        return mac.doFinal(data)
    }

    companion object {
        private const val HMAC_ALG = "HmacSHA256"
        private const val SHA256_LEN = 32
    }
}

File: metrics/codegen/src/main/resources/swift/MetricsStableIdEncoder.swift
import CryptoKit
import Foundation

public struct MetricsStableIdEncoder {
    public let salt: Data
    public let info: Data

    public init(salt: Data, info: Data) {
        self.salt = salt
        self.info = info
    }

    public func derive(privateKey: Data) -> String {
        let key = HKDF<SHA256>.deriveKey(
            inputKeyMaterial: SymmetricKey(data: privateKey),
            salt: salt,
            info: info,
            outputByteCount: 32
        )
        return key.withUnsafeBytes { Data($0) }
            .map { String(format: "%02x", $0) }
            .joined()
    }
}
